src/corefx/System.Globalization.Native/pal_collation.c

```diff
@@ -385,38 +377,36 @@ void GlobalizationNative_CloseSortHandle(SortHandle* pSortHandle)
         }
     }
 
-    pthread_mutex_destroy(&pSortHandle->collatorsLockObject);
-
     free(pSortHandle);
 }
 
 const UCollator* GetCollatorFromSortHandle(SortHandle* pSortHandle, int32_t options, UErrorCode* pErr)
 {
-    UCollator* pCollator;
     if (options == 0)
     {
-        pCollator = pSortHandle->collatorsPerOption[0];
+        return pSortHandle->collatorsPerOption[0];
     }
     else
     {
-        int lockResult = pthread_mutex_lock(&pSortHandle->collatorsLockObject);
-        if (lockResult != 0)
+        options &= CompareOptionsMask;
+        UCollator* pCollator = pSortHandle->collatorsPerOption[options];
+        if (pCollator != NULL)
         {
-            assert(FALSE && "Unexpected pthread_mutex_lock return value.");
+            return pCollator;
         }
 
-        options &= CompareOptionsMask;
-        pCollator = pSortHandle->collatorsPerOption[options];
-        if (pCollator == NULL)
+        pCollator = CloneCollatorWithOptions(pSortHandle->collatorsPerOption[0], options, pErr);
+
+        UCollator* pNull = NULL; // callee requires a non-null argument
+        if (!atomic_compare_exchange_strong(&pSortHandle->collatorsPerOption[options], &pNull, pCollator))
```

Comment:
>&pNull [](start = 87, length = 6)

this should be pNull and nor &pNull

Reply:
maybe I am wrong here :-) anyway I hope we have tested it.

---
In reply to: [291326461](https://github.com/dotnet/coreclr/pull/24973#discussion_r291326461) [](ancestors = 291326461)

Reply:
If I change it to `pNull` it fails to compile:

```log
/home/adsitnik/Projects/coreclr/src/corefx/System.Globalization.Native/pal_collation.c:401:88: error: incompatible pointer types passing 'UCollator *' (aka 'struct UCollator *') to parameter of type 'UCollator **'
      (aka 'struct UCollator **'); take the address with & [-Werror,-Wincompatible-pointer-types]
/home/adsitnik/Projects/coreclr/src/corefx/System.Globalization.Native/pal_collation.c:        if (!atomic_compare_exchange_strong(&pSortHandle->collatorsPerOption[options], pNull, pCollator))401
                                                                                       ^~~~~
                                                                                       &
:/usr/lib/llvm-3.9/bin/../lib/clang/3.9.1/include/stdatomic.h:140:112: 88note::  expanded from macro 'atomic_compare_exchange_strong'
error: #define atomic_compare_exchange_strong(object, expected, desired) __c11_atomic_compare_exchange_strong(object, expected, desired, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST)
                                                                                                               ^~~~~~~~
incompatible pointer types passing 'UCollator *' (aka 'struct UCollator *') to parameter of type 'UCollator **'
      (aka 'struct UCollator **'); take the address with & [-Werror,-Wincompatible-pointer-types]
        if (!atomic_compare_exchange_strong(&pSortHandle->collatorsPerOption[options], pNull, pCollator))
                                                                                       ^~~~~
                                                                                       &
/usr/lib/llvm-3.9/bin/../lib/clang/3.9.1/include/stdatomic.h:140:112: note: expanded from macro 'atomic_compare_exchange_strong'
#define atomic_compare_exchange_strong(object, expected, desired) __c11_atomic_compare_exchange_strong(object, expected, desired, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST)
```

If I change it to `NULL` it compiles (on my vm, not on the CI :) ) but then the program fails with:

```log
Illegal instruction (core dumped)
```